rotors_gazebo_plugins/src/gazebo_wind_plugin.cpp

```diff
@@ -99,38 +169,160 @@ void GazeboWindPlugin::OnUpdate(const common::UpdateInfo& _info) {
+          wind_field_cells[0][3].z = coords_z_0[i+1];
+        }
+      for (int i = 0; i < coords_z_1.size(); i++)
+        if (coords_z_1[i] < link_position.z && coords_z_1[i+1] > link_position.z)
+        {
+          wind_field_cells[0][1].z = coords_z_1[i];
+          wind_field_cells[0][2].z = coords_z_1[i+1];
+        }
+      for (int i = 0; i < coords_z_4.size(); i++)
+        if (coords_z_4[i] < link_position.z && coords_z_4[i+1] > link_position.z)
+        {
+          wind_field_cells[0][4].z = coords_z_4[i];
+          wind_field_cells[0][7].z = coords_z_4[i+1];
+        }
+      for (int i = 0; i < coords_z_5.size(); i++)
+        if (coords_z_5[i] < link_position.z && coords_z_5[i+1] > link_position.z)
+        {
+          wind_field_cells[0][5].z = coords_z_5[i];
+          wind_field_cells[0][6].z = coords_z_5[i+1];
+        }
+
+      wind_field_cells[0][8].z =  wind_field_cells[0][9].z =  wind_field_cells[0][10].z =  wind_field_cells[0][11].z =  wind_field_cells[0][12].z =  wind_field_cells[0][13].z = link_position.z;
+
+      // Extract velocity components of each of the 8 given vertices.
+      for (int i = 0; i < wind_field_[1].size(); i++)
+        for (int j = 0; j < 8; j++)
+          if (wind_field_[0][i] == wind_field_cells[0][j])
+              wind_field_cells[1][j] = wind_field_[1][i];
+
+      // Interpolate three times linearly to find the value of the wind velocity at the given position
+      wind_field_cells[1][8] = wind_field_cells[1][0] + (wind_field_cells[1][3] - wind_field_cells[1][0])/(wind_field_cells[0][3].z - wind_field_cells[0][0].z)*(wind_field_cells[0][8].z - wind_field_cells[0][0].z);
+      wind_field_cells[1][9] = wind_field_cells[1][1] + (wind_field_cells[1][2] - wind_field_cells[1][1])/(wind_field_cells[0][2].z - wind_field_cells[0][1].z)*(wind_field_cells[0][9].z - wind_field_cells[0][1].z);
+      wind_field_cells[1][10] = wind_field_cells[1][5] + (wind_field_cells[1][6] - wind_field_cells[1][5])/(wind_field_cells[0][6].z - wind_field_cells[0][5].z)*(wind_field_cells[0][10].z - wind_field_cells[0][5].z);
+      wind_field_cells[1][11] = wind_field_cells[1][4] + (wind_field_cells[1][7] - wind_field_cells[1][4])/(wind_field_cells[0][7].z - wind_field_cells[0][4].z)*(wind_field_cells[0][11].z - wind_field_cells[0][4].z);
+
+      wind_field_cells[1][12] = wind_field_cells[1][8] + (wind_field_cells[1][11] - wind_field_cells[1][8])/(wind_field_cells[0][11].y - wind_field_cells[0][8].y)*(wind_field_cells[0][12].y - wind_field_cells[0][8].y);
+      wind_field_cells[1][13] = wind_field_cells[1][9] + (wind_field_cells[1][10] - wind_field_cells[1][9])/(wind_field_cells[0][10].y - wind_field_cells[0][9].y)*(wind_field_cells[0][13].y - wind_field_cells[0][9].y);
+
+      wind_velocity = wind_field_cells[1][12] + (wind_field_cells[1][13] - wind_field_cells[1][12])/(wind_field_cells[0][13].x - wind_field_cells[0][12].x)*(link_position.x - wind_field_cells[0][12].x);
```

Comment:
General comment: Did you write a unit test for this? I.e. is all this verified? I assume it is extremely easy to do a small coding mistake here that you will otherwise not find.